src/cli/commands/upgrade.js

```diff
@@ -21,12 +34,95 @@ export function hasWrapper(commander: Object, args: Array<string>): boolean {
 export const requireLockfile = true;
 
 export async function run(config: Config, reporter: Reporter, flags: Object, args: Array<string>): Promise<void> {
-  const useLockfile = args.length || flags.latest;
-  const lockfile = useLockfile ? await Lockfile.fromDirectory(config.lockfileFolder, reporter) : new Lockfile();
-  const {dependencies, devDependencies, optionalDependencies, peerDependencies} =
-    (await config.readRootManifest()) || {};
-  const allDependencies = Object.assign({}, peerDependencies, optionalDependencies, devDependencies, dependencies);
-  let addArgs = [];
+  const lockfile = await Lockfile.fromDirectory(config.lockfileFolder);
+  const deps = await getOutdated(config, reporter, flags, lockfile, args);
+
+  // if specific versions were requested for packages, override what getOutdated reported as the latest to install
+  args.forEach(requestedPattern => {
+    const normalized = PackageRequest.normalizePattern(requestedPattern);
+    const newPattern = `${normalized.name}@${normalized.range}`;
+    let found = false;
+
+    deps.forEach(dep => {
+      if (normalized.hasVersion && dep.name === normalized.name) {
+        found = true;
+        dep.latestPattern = newPattern;
+      }
+    });
+
+    if (normalized.hasVersion && !found) {
+      deps.push({
+        name: normalized.name,
+        wanted: '',
+        latest: '',
+        url: '',
+        hint: '',
+        range: '',
+        current: '',
+        latestPattern: newPattern,
+      });
+    }
+  });
+
+  if (!deps.length) {
+    reporter.success(reporter.lang('allDependenciesUpToDate'));
+    return;
+  }
+
+  // remove deps being upgraded from the lockfile, or else Add will use the already-installed version
+  // instead of the latest for the range.
+  deps.forEach(dep => lockfile.removePattern(dep.latestPattern));
+
+  const addFlags = Object.assign({}, flags, {force: true});
```

Comment:
Maybe. Probably. This is what the old/existing code did so I preserved it without much thought. Does the `commander` library that handles the command line options have anything to enforce only the allowed flags?